### ISSUE TITLE
Validation pattern accessibility requirement (potentially breaking change)

### ISSUE THREAD

**Issue by lyubomir-popov** [canonical-web-and-design/vanilla-framework#3324]
The following seem to require modifying labels, and placing links leading to fields that fail validation _under_ the submit button. This would be a pattern change, probably would require UX input too. Flagging for discussion.

Sources: 
https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=413#status-messages
https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html
![image](https://user-images.githubusercontent.com/2741678/94566225-03477f80-0262-11eb-8deb-5f1b7959804b.png)

**Comment by anthonydillon** [comment]
This sounds like two changes:
1. Add a change to the label when field is in error. This will require UX/design but should not be a breaking change.
2. Add JS documentation to inject links below the button on failed submittions.

**Comment by matthewpaulthomas** [comment]
Have we seen any examples of sites that do either of those things? Both seem quite strange.

> seem to require

“[Techniques are informative—that means they are not required.](https://www.w3.org/WAI/WCAG21/Understanding/understanding-techniques#techniques-are-informative)”

“[the contents of this document … do not set requirements for conforming to WCAG 2.1](https://www.w3.org/WAI/WCAG21/Understanding/#abstract)”

> 1. Add a change to the label when field is in error.

If this means adding it to the `<label>`, I guess the error message could still be positioned separately, something like:
```html
<label>Password:
  <input type="password">
  <span role="alert" aria-atomic="true">Too short</span>
</label>
```

> 2. Add JS […] to inject links below the button on failed submissions.

Benefit: If the form is shorter than the page, the errors would appear immediately below where you’d just clicked, without any scrolling or jumping of the form.

Drawbacks:
- If you’d scrolled down just far enough to see the submit button (for example, by Tabbing to it), the links would be off-screen.
- Inconsistent with forms that have a traditional reload-on-submit, where errors are therefore summarized above the form.

**Comment by lyubomir-popov** [comment]
I think it means adding a link (not necessarily a visible one) under the button, so that a screen reader (all of this is for blind users only) can go back to the field that errored out. I guess the intention is to not have to ask a blind user to listen to the screen reader narrate the entire page again just to get to the one field that produced a validation error.

I'm by no means advocating for it, just trying to interpret what these guidelines mean. Especially given we can't really test this. I tried Mac OS voiceOver and it just reads the source code...

**Comment by lyubomir-popov** [comment]
To be looked at as part of an external accessibility as we have no means of testing and we haven't seen anything like that done anywhere.

**Comment by bartaz** [comment]
TODO:

- Update validation pattern to distinguish labels in fields with errors (UX/design needed)
- Experiment with screen reader and invisible links under the button

**Comment by bartaz** [comment]
Blocked. We will get back to it when audit happens.

**Comment by clagom** [comment]
@wgx and @AnnSoMuller will consider this issue when working on #3766

**Comment by wgx** [comment]
Closing, resolved by work here: https://github.com/canonical-web-and-design/vanilla-framework/issues/3766
